Foundation/NSURLSession/http/HTTPURLProtocol.swift

```diff
@@ -0,0 +1,892 @@
+        easyHandle.set(skipAllSignalHandling: true)
+
+        // Error Options:
+        easyHandle.set(errorBuffer: nil)
+        easyHandle.set(failOnHTTPErrorCode: false)
+
+        // Network Options:
+        guard let url = request.url else { fatalError("No URL in request.") }
+        easyHandle.set(url: url)
+        easyHandle.setAllowedProtocolsToHTTPAndHTTPS()
+        easyHandle.set(preferredReceiveBufferSize: Int.max)
+        do {
+            switch (task?.body, try task?.body.getBodyLength()) {
+            case (.none, _):
+                set(requestBodyLength: .noBody)
+            case (_, .some(let length)):
+                set(requestBodyLength: .length(length))
+            case (_, .none):
+                set(requestBodyLength: .unknown)
+            }
+        } catch let e {
+            // Fail the request here.
+            // TODO: We have multiple options:
+            //     NSURLErrorNoPermissionsToReadFile
+            //     NSURLErrorFileDoesNotExist
+            self.internalState = .transferFailed
+            failWith(errorCode: errorCode(fileSystemError: e), request: request)
+            return
+        }
+
+        // HTTP Options:
+        easyHandle.set(followLocation: false)
+
+        // The httpAdditionalHeaders from session configuration has to be added to the request.
+        // The request.allHTTPHeaders can override the httpAdditionalHeaders elements. Add the
+        // httpAdditionalHeaders from session configuration first and then append/update the
+        // request.allHTTPHeaders so that request.allHTTPHeaders can override httpAdditionalHeaders.
+
+        let httpSession = self.task?.session as! URLSession
+        var httpHeaders: [AnyHashable : Any]?
+
+        if let hh = httpSession.configuration.httpAdditionalHeaders {
+            httpHeaders = hh
+        }
+
+        if let hh = self.task?.originalRequest?.allHTTPHeaderFields {
+            if httpHeaders == nil {
+                httpHeaders = hh
+            } else {
+                hh.forEach {
+                    httpHeaders![$0] = $1
+                }
+            }
+        }
+        let customHeaders: [String]
+        let headersForRequest = curlHeaders(for: httpHeaders)
+        if ((request.httpMethod == "POST") && (request.value(forHTTPHeaderField: "Content-Type") == nil)) {
+            customHeaders = headersForRequest + ["Content-Type:application/x-www-form-urlencoded"]
+        } else {
+            customHeaders = headersForRequest
+        }
+
+        easyHandle.set(customHeaders: customHeaders)
+
+        //TODO: The CURLOPT_PIPEDWAIT option is unavailable on Ubuntu 14.04 (libcurl 7.36)
+        //TODO: Introduce something like an #if, if we want to set them here
+
+        //set the request timeout
+        //TODO: the timeout value needs to be reset on every data transfer
+
+        var timeoutInterval = Int(httpSession.configuration.timeoutIntervalForRequest) * 1000
```

Comment:
This timeout logic needs to be moved to URLSessionTask since it is not something specific to HTTP. We can do this in a separate PR though.